tests/test_autosave.py

```diff
@@ -0,0 +1,492 @@
+    select_and_recv(parent_conn, "D")
+
+
+def check_autosave_field_names_contain_device_prefix(
+    device_name, tmp_path, conn
+):
+    autosave.configure(tmp_path, device_name, save_period=1)
+    builder.aOut("BEFORE", autosave=["VAL", "EGU"])
+    builder.SetDeviceName(device_name)
+    builder.aOut("AFTER", autosave=["VAL", "EGU"])
+    builder.LoadDatabase()
+    softioc.iocInit()
+    time.sleep(2)
+    with open(tmp_path / f"{device_name}.softsav", "r") as f:
+        saved = yaml.full_load(f)
+    assert "BEFORE" in saved.keys()
+    assert f"{device_name}:AFTER" in saved.keys()
+    autosave.Autosave._stop()
+    conn.send("D")
+
+
+def test_autosave_field_names_contain_device_prefix(tmp_path):
+    ctx = get_multiprocessing_context()
+    parent_conn, child_conn = ctx.Pipe()
+    ioc_process = ctx.Process(
+        target=check_autosave_field_names_contain_device_prefix,
+        args=(DEVICE_NAME, tmp_path, child_conn),
+    )
+    ioc_process.start()
+    # If we never receive D it probably means an assert failed
+    select_and_recv(parent_conn, "D")
+
+def test_context_manager_thread_safety(tmp_path):
+    autosave.configure(tmp_path, DEVICE_NAME)
+
+    def create_pv_in_thread(name, wait):
+        time.sleep(wait)
+        builder.aOut(name, autosave=False)
+
+    pv_thread_before_cm = threading.Thread(
+        target=create_pv_in_thread, args=["PV-FROM-THREAD-BEFORE", 1])
+    pv_thread_in_cm = threading.Thread(
+        target=create_pv_in_thread, args=["PV-FROM-THREAD-DURING", 0])
+    pv_thread_before_cm.start()
+    with autosave.Autosave(["VAL", "EGU"]):
+        builder.aOut("PV-FROM-CM")
+        pv_thread_in_cm.start()
+        pv_thread_in_cm.join()
+    pv_thread_before_cm.join()
+
+    assert "PV-FROM-THREAD-BEFORE" not in autosave.Autosave._pvs
+    assert "PV-FROM-THREAD-DURING" not in autosave.Autosave._pvs
+    assert device_core.LookupRecord("PV-FROM-THREAD-BEFORE")
+    assert device_core.LookupRecord("PV-FROM-THREAD-DURING")
+
+def test_nested_context_managers_raises(tmp_path):
+    autosave.configure(tmp_path, DEVICE_NAME)
+    with autosave.Autosave(["SCAN"]):
+        with pytest.raises(RuntimeError):
+            with autosave.Autosave(False):
+                builder.aOut("MY-PV")
+        with pytest.raises(RuntimeError):
+            autosave.Autosave()
+
+def test_autosave_arguments(tmp_path):
+    autosave.configure(tmp_path, DEVICE_NAME)
+    builder.aOut("TRUE", autosave=True)
+    builder.aOut("FIELDS", autosave=["LOPR", "HOPR"])
+    builder.aOut("FALSE", autosave=False)
+    builder.aOut("SINGLE-FIELD", autosave="EGU")
+    builder.aOut("AUTO-VAL", autosave="VAL")
+    assert set(autosave.Autosave._pvs) == {
+        "TRUE", "FIELDS.LOPR", "FIELDS.HOPR", "SINGLE-FIELD.EGU", "AUTO-VAL"}
+    autosave.Autosave._pvs = {}
+    builder.ClearRecords()
+    with autosave.Autosave():  # True by default
+        builder.aOut("AUTO-TRUE", autosave=False)
```

Comment:
Something seems wrong here; specifying `autosave=False` is ignored (as it looks like this PV does appear in the saved PV list)? That's a different behaviour to below (lines 486-487) where specifying `autosave=True` does override the CM's False flag.

Reply:
Okay, setting autosave=False explicitly disables autosave for the pv inside a context manager now.